### PR TITLE
Drop deprecated pointer package

### PR DESCRIPTION
**Pull request by liyuerich**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Use functions in k8s.io/utils/ptr instead:  ptr.Deref to dereference a pointer

#### Which issue(s) this PR fixes:

Ref # https://github.com/kubernetes/utils/pull/283

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```
NONE
```

### CODE REVIEW

**Comment by liyuerich** [comment]
> Hello. I'd suggest opening up smaller PRs that span a single SIG. It is hard to get reviews accross sigs and it would maybe go faster if you split these up.
> 
> As you see, you'll need at least 8 approvers for this PR.

thanks @kannon92 , will also combine some together due to only 1 or 2 files are changed. do you think it  makes sense?

/cc @carlory  @skitt

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -179,8 +179,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
...
 	fs.Int32Var(&o.metricsPort, "metrics-port", o.metricsPort, "The port to bind the metrics server. Use 0 to disable.")
 	fs.MarkDeprecated("metrics-port", "This flag is deprecated and will be removed in a future release. Please use --metrics-bind-address instead.")
-	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", pointer.Int32Deref(o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
-	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", pointer.Int32Deref(o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
+	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", ptr.Deref[int32](o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:182]: The type can be inferred:

```suggestion
	fs.Int32Var(o.config.OOMScoreAdj, "oom-score-adj", ptr.Deref(o.config.OOMScoreAdj, int32(qos.KubeProxyOOMScoreAdj)), "The oom-score-adj value for kube-proxy process. Values must be within the range [-1000, 1000]. This parameter is ignored if a config file is specified by --config.")
```

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -179,8 +179,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
...
+	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", ptr.Deref[int32](o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:183]: Same here:

```suggestion
	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", ptr.Deref(o.config.IPTables.MasqueradeBit, 14), "If using the pure iptables proxy, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
```

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -203,7 +203,7 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.config.IPVS.StrictARP, "ipvs-strict-arp", o.config.IPVS.StrictARP, "Enable strict ARP by setting arp_ignore to 1 and arp_announce to 2")
 	fs.BoolVar(&o.config.IPTables.MasqueradeAll, "masquerade-all", o.config.IPTables.MasqueradeAll, "If using the pure iptables proxy, SNAT all traffic sent via Service cluster IPs (this not commonly needed)")
-	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", pointer.BoolDeref(o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
+	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", ptr.Deref[bool](o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
```
Comment (skitt) [inline cmd/kube-proxy/app/server.go:206]: Same here:

```suggestion
	fs.BoolVar(o.config.IPTables.LocalhostNodePorts, "iptables-localhost-nodeports", ptr.Deref(o.config.IPTables.LocalhostNodePorts, true), "If false Kube-proxy will disable the legacy behavior of allowing NodePort services to be accessed via localhost, This only applies to iptables mode and ipv4.")
```

**Review by skitt — COMMENTED** [review]
Most of the time the type argument to `ptr.Deref` can be inferred; I highlighted the first three, but please remove all those that can be removed.

**Comment by liyuerich** [comment]
> > > Hello. I'd suggest opening up smaller PRs that span a single SIG. It is hard to get reviews accross sigs and it would maybe go faster if you split these up.
> > > As you see, you'll need at least 8 approvers for this PR.
> > 
> > 
> > thanks @kannon92 , will also combine some together due to only 1 or 2 files are changed. do you think it makes sense?
> > /cc @carlory @skitt
> 
> Natural separation for this could be:
> 
> PR for things in pkg/controllers. PR for kube-proxy PR for staging changes and API changes

@kannon92 , after check,  I find there is only 1 file changed in kube-proxy and there are 7 files changed in pkg/controllers.  It might combine pkg/controllers & kube-proxy together. is it ok?

**Comment by liyuerich** [comment]
it seems the split operation reverts all my changes "type when it can be inferred." I will update again

File: pkg/api/job/warnings.go [hunk 4]
```diff
@@ -36,8 +36,8 @@ const (
 func WarningsForJobSpec(ctx context.Context, path *field.Path, spec, oldSpec *batch.JobSpec) []string {
 	var warnings []string
 	if spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion {
-		completions := pointer.Int32Deref(spec.Completions, 0)
-		parallelism := pointer.Int32Deref(spec.Parallelism, 0)
+		completions := ptr.Deref[int32](spec.Completions, 0)
+		parallelism := ptr.Deref[int32](spec.Parallelism, 0)
```
Comment (carlory) [inline pkg/api/job/warnings.go:40]: Can we drop `[int32]`?

Comment (liyuerich) [reply pkg/api/job/warnings.go]: sure. thanks

**Comment by liyuerich** [comment]
hello,
@liggitt @smarterclayton could you help to review since there is lgtm label? thanks.

**Comment by liyuerich** [comment]
fix the conflict , please review again. thanks

**Pull request merged**